portal-service/src/com/liferay/portal/model/PortletInstance.java

```diff
@@ -114,6 +133,48 @@ public String toString() {
 		return getPortletInstanceKey();
 	}
 
+	private static FullInstanceId _buildFullInstanceId(String fullInstanceId) {
+		if (fullInstanceId == null) {
+			throw new InvalidParameterException(
+				"The fullInstanceId must not be null.");
+		}
+
+		String[] tokens = StringUtil.split(
+			fullInstanceId, StringPool.UNDERLINE);
+
+		int count = StringUtil.count(fullInstanceId, StringPool.UNDERLINE);
+
+		if ((tokens.length > 2) || (count > 1)) {
+			throw new InvalidParameterException(
+				"The fullInstanceId must only contain one underscore " +
+					"separating the instanceId from the userId.");
+		}
+
+		if (tokens.length == 2) {
+			long userId = GetterUtil.getLong(tokens[0], -1);
+
+			if (userId == -1) {
+				throw new InvalidParameterException(
+					"The fullInstanceId's userId must be a valid number.");
+			}
+
+			return new FullInstanceId(tokens[1], userId);
+		}
+		else if (tokens.length == 1) {
+			int index = fullInstanceId.indexOf(StringPool.UNDERLINE);
```

Comment:
I find this mix of tokenization and indexOf confusing. If the tokenization is not really enough I tend to think that it's better to just do indexOf and substrings based on that. In any case it might be a matter of taste.